File: core/src/com/qtech/bubbleblaster/common/logging/GameLogger.java
package com.qtech.bubbleblaster.common.logging;

import com.qtech.bubbleblaster.common.References;

import javax.management.InstanceAlreadyExistsException;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Logger;

public class GameLogger {
    private final Logger logger;
    private static final ConcurrentHashMap<String, GameLogger> loggers = new ConcurrentHashMap<>();
    private final GameLogFormatter formatter = new GameLogFormatter();
    private String name;

    public void debug(String msg) {
        this.log(GameLogLevel.DEBUG, msg);
    }

    public void info(String msg) {
        this.log(GameLogLevel.INFO, msg);
    }

    public void success(String msg) {
        this.log(GameLogLevel.SUCCESS, msg);
    }

    public void error(String msg) {
        this.log(GameLogLevel.ERROR, msg);
    }

    public void warning(String msg) {
        this.log(GameLogLevel.WARN, msg);
    }

    public void fatal(String msg) {
        this.log(GameLogLevel.FATAL, msg);
    }

    private void log(GameLogLevel level, String msg) {
        GameLogRecord record = new GameLogRecord(msg, this, level);
        this.formatter.publish(record, printer);
    }

    private static PrintWriter printer = null;

    static {
        try {
            Date date = new Date(System.currentTimeMillis());
            DateFormat formatter = new SimpleDateFormat("dd_MM_yyyy HH_mm_ss");
            formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
            String logFileName = formatter.format(date) + ".log";

            FileOutputStream output = new FileOutputStream(new File(References.LOGS_DIR, logFileName), true);
            printer = new PrintWriter(output);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public GameLogger(String name) throws InstanceAlreadyExistsException {
        if (loggers.containsKey(name))
            throw new InstanceAlreadyExistsException("GameLogger instance with name '" + name + "' already created");

        // Get Java logger.
        this.logger = Logger.getLogger(name);
        this.name = name;

        // Register logger.
        loggers.put(name, this);
    }

    public static ConcurrentHashMap<String, GameLogger> getLoggers() {
        return loggers;
    }

    public static GameLogger getLoggerInstance(String name) {
        return loggers.get(name);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Logger getLogger() {
        return logger;
    }
}

File: src/main/kotlin/qtech/bubbles/common/logging/GameLogger.kt
@file:Suppress("unused")

package qtech.bubbles.common.logging

import qtech.bubbles.common.References
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.PrintWriter
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.logging.Logger
import javax.management.InstanceAlreadyExistsException

class GameLogger(name: String) {
    val logger: Logger
    private val formatter = GameLogFormatter()
    var name: String
    fun debug(msg: String) {
        log(GameLogLevel.DEBUG, msg)
    }

    fun info(msg: String) {
        log(GameLogLevel.INFO, msg)
    }

    fun success(msg: String) {
        log(GameLogLevel.SUCCESS, msg)
    }

    fun error(msg: String) {
        log(GameLogLevel.ERROR, msg)
    }

    fun warning(msg: String) {
        log(GameLogLevel.WARN, msg)
    }

    fun fatal(msg: String) {
        log(GameLogLevel.FATAL, msg)
    }

    private fun log(level: GameLogLevel, msg: String) {
        val record = GameLogRecord(msg, this, level)
        formatter.publish(record, printer)
    }

    companion object {
        val loggers = ConcurrentHashMap<String, GameLogger>()
        private var printer: PrintWriter? = null
        fun getLoggerInstance(name: String): GameLogger? {
            return loggers[name]
        }

        init {
            try {
                val date = Date(System.currentTimeMillis())
                val formatter: DateFormat = SimpleDateFormat("dd_MM_yyyy HH_mm_ss")
                formatter.timeZone = TimeZone.getTimeZone("UTC")
                val logFileName: String = formatter.format(date) + ".log"
                val output = FileOutputStream(File(References.LOGS_DIR, logFileName), true)
                printer = PrintWriter(output)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    init {
        if (loggers.containsKey(name)) throw InstanceAlreadyExistsException("GameLogger instance with name '$name' already created")

        // Get Java logger.
        logger = Logger.getLogger(name)
        this.name = name

        // Register logger.
        loggers[name] = this
    }
}